Automattic/vip-go-mu-plugins
# Add VIP Integrations framework

## Description

The previous integration was removed in [this PR](https://github.com/Automattic/vip-go-mu-plugins/pull/4451) to re-think the approach used to enable plugins. It's now up again with the logic being:

- Call the higher level method `load_integration` and pass in the correct slug along with any config options that are supported.

This makes it extendable in the future for other integrations, as well as provides the ability to change the way integrations are enabled as their implementation is hidden away.

In addition to that, the vendor folder for the block-data-api is now ignored by the linter. Without this, the linter would fail. Previously, the `dockerignore` file was updated with files to ignore. This reverses that, and use the lint ignore file instead.

There is a corresponding PR to reverse that dockerignore fix in the mu-plugins-ext: https://github.com/Automattic/vip-go-mu-plugins-ext/pull/35

This PR and the above PR should be merged at the same time to avoid build failures.

## Changelog Description

### Added high level method to enable VIP Integration Plugins

Added support to enable VIP Integration Plugins, starting with the [VIP Block Data API plugin](https://github.com/Automattic/vip-block-data-api/) via the `load_integration` method.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] N/A - This change uses a rollout method to ease with deployment.
- [x] External documentation in progress - This change has relevant documentation additions / updates (if applicable).

## Steps to Test

Ensure you have [this PR's change](https://github.com/Automattic/vip-go-mu-plugins-ext/pull/35) in your vip-go-mu-plugins-ext.

1. Wait until a `vip-go-mu-plugins` deployment is made, which should pull [the `vip-integrations` folder from `-ext`](https://github.com/Automattic/vip-go-mu-plugins-ext/tree/trunk/vip-integrations) into [vip-go-mu-plugins-built](https://github.com/Automattic/vip-go-mu-plugins-built).
2. Clone [vip-go-mu-plugins-built](https://github.com/Automattic/vip-go-mu-plugins-built) locally set `vip dev-env`'s environment to use it, e.g.

    ```bash
    vip dev-env update --mu-plugins=~/projects/mu-plugins-built
    ```

3. Verify the block data API plugin is disabled by default. Visit the REST endpoint for an existing post ID:

    ```bash
    GET /wp-json/vip-block-data-api/v1/posts/<post_id>/blocks

    # e.g. GET http://my-site.vipdev.lndo.site/wp-json/vip-block-data-api/v1/posts/1/blocks
    ```

    Ensure you receive a `404`response with status code `rest_no_route`.

4. In `client-mu-plugins/plugin-loader.php`, add:

    ```php
    \Automattic\VIP\Integrations\activate( 'block-data-api' );
    ```

6. Visit the same URL as step 3, and ensure the block data API is now available:

    ```
    GET http://my-site.vipdev.lndo.site/wp-json/vip-block-data-api/v1/posts/1/blocks

    {"blocks":[{"name":"core/paragraph","attributes":{"content":"Welcome to WordPress. This is your first post. Edit or delete it, then start writing!","dropCap":false}}]}
    ```

---

### Unit tests

Run:

```
./bin/test.sh --filter VIP_Integrations_Test
```
